snap/snapshots.go

```diff
@@ -27,27 +27,104 @@ import (
 	"path/filepath"
 	"strings"
 
-	"gopkg.in/yaml.v2"
-
 	"github.com/snapcore/snapd/strutil"
+	"gopkg.in/yaml.v2"
 )
 
 var osOpen = os.Open
 
-// SnapshotOptions contains the information provided by a snap package about
-// what files should be included in the snapshots.
+// SnapshotOptions describes the options available for snapshots.
+// The initial source of these options is a file in the snap package.
+// In addition, options can be modified with dynamic requests via REST API.
 type SnapshotOptions struct {
 	// ExcludePaths is the list of file and directory patterns that need to be
 	// excluded from a snapshot. At the moment the only supported globbing
 	// character is "*", which stands for any sequence of characters other than
 	// "/".
-	ExcludePaths []string `yaml:"exclude"`
+	ExcludePaths []string `yaml:"exclude" json:"exclude,omitempty"`
 }
 
 const (
 	snapshotManifestPath = "meta/snapshots.yaml"
 )
 
+func (opts *SnapshotOptions) validateExcludePaths() error {
+	if err := validateExcludePaths(opts.ExcludePaths); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (opts *SnapshotOptions) mergeExcludePaths(paths []string) error {
+	if err := validateExcludePaths(paths); err != nil {
+		return err
+	}
+	opts.ExcludePaths = append(opts.ExcludePaths, paths...)
+
+	return nil
+}
+
+// IsEmpty determines if SnapshotOptions structure is empty.
+func (opts *SnapshotOptions) IsEmpty() bool {
```

Comment:
The reason for this "empty" check is because un-marshalling of some variation of empty options struct e.g. `"options": {"snapname":{}}` would result in a mapping to an empty slice, even when using map to pointers of options. In these cases the slice is not nil, and in my understanding the "unset" state would typically be understood to be nil?

A generic `reflect.ValueOf(options).IsZero()` does not work to detect this state

So in this context, my proposal is that perhaps empty is more suitable?
 - detect condition on which we want to intervene to have `json: "..., omitempty"` leave out the structure
 - condition is not detecting nest "unset" containers, but rather nest of containers without values

I modified the comment to try and explain what is meant by "empty". Let me know if you still feel this is not the right way?

Reply:
We use `Unset` sometimes like golangs normal naming convention `IsZero`, it's meant to convey that the value is in its default state. 

In any case I don't think `IsEmpty` is correctly conveying intent here, because from a code perspective `Unset` or `IsZero` clearly tells us that either the structure is unset, or the structure is zero (default values), while `IsEmpty` would be more an operation on a container, and doesn't make the same sense in the case of a non-container

Reply:
Changed name to Unset as per discussion.